### PR TITLE
Add GPU clock policy compliance health check

### PR DESCRIPTION
**Pull request by theap06**
Summary
Add a new GPU health check to validate application clock policy compliance and detect per-GPU clock drift.

This introduces a dedicated command, policy schema/evaluator, CLI wiring, tests, and documentation.

Motivation
Clock misconfiguration can silently degrade GPU workload performance.
This check provides an explicit policy gate for expected GPU application clocks and reports drift severity using existing health check exit semantics.

Test Plan:

Ran targeted tests for the new GPU clock policy check:

pytest -q [test_application_clock_policy.py](http://_vscodecontentref_/0) gcm/tests/health_checks_tests/test_gpu_clock_policy.py
Result:

9 passed in 0.33s

python3 -m gcm.health_checks.cli.health_checks \
  check-gpu-clock-policy \
  lambda-test \
  nagios \
  --sink stdout \
  --expected-graphics-freq 1695 \
  --expected-memory-freq 6251
  
 Output:
OK - nvidia smi clock policy
[{"node": "129-159-36-52", "gpu_node_id": "ff2ef02bb225cce9878e95fc12f4e3e39aef0fcff8d5de2a6c16a900ed763e66", "cluster": "lambda-test", "derived_cluster": "lambda-test", "health_check": "nvidia smi clock policy", "type": "nagios", "result": 0, "_msg": "clock_policy check: GPU 0, severity=OK, expected=(graphics:1695, memory:6251), observed=(graphics:1695, memory:6251), delta_mhz=(graphics:0, memory:0)\n", "job_id": 0, "start_time": 1772265789.8423011, "end_time": 1772265789.8663087}]

echo $?
output:
0

python3 -m gcm.health_checks.cli.health_checks \
  check-gpu-clock-policy \
  lambda-test \
  nagios \
  --sink stdout \
  --expected-graphics-freq 9999 \
  --expected-memory-freq 9999
  
 output:
CRITICAL - nvidia smi clock policy
[{"node": "129-159-36-52", "gpu_node_id": "ff2ef02bb225cce9878e95fc12f4e3e39aef0fcff8d5de2a6c16a900ed763e66", "cluster": "lambda-test", "derived_cluster": "lambda-test", "health_check": "nvidia smi clock policy", "type": "nagios", "result": 2, "_msg": "clock_policy check: GPU 0, severity=CRITICAL, expected=(graphics:9999, memory:9999), observed=(graphics:1695, memory:6251), delta_mhz=(graphics:8304, memory:3748)\n", "job_id": 0, "start_time": 1772265802.1543877, "end_time": 1772265802.1807373}]

ubuntu@129-159-36-52:~/gcm$ echo $?

output:
2

### CODE REVIEW

File: gcm/health_checks/checks/check_gpu_clock_policy.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+def severity_to_exitcode(severity: ClockComplianceSeverity) -> ExitCode:
+    if severity == ClockComplianceSeverity.CRITICAL:
+        return ExitCode.CRITICAL
+    if severity == ClockComplianceSeverity.WARN:
+        return ExitCode.WARN
+    return ExitCode.OK
+
+
+@click.command()
+@common_arguments
```
Comment (luccabb) [inline gcm/health_checks/checks/check_gpu_clock_policy.py]:
```suggestion
@common_arguments
@timeout_argument
```

File: gcm/schemas/gpu/application_clock_policy.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def evaluate_clock_policy(
+    observed: ApplicationClockInfo, policy: ClockPolicy
+) -> ClockComplianceResult:
+    graphics_delta_mhz = abs(observed.graphics_freq - policy.expected_graphics_freq)
+    memory_delta_mhz = abs(observed.memory_freq - policy.expected_memory_freq)
+    max_delta_mhz = max(graphics_delta_mhz, memory_delta_mhz)
+
+    if max_delta_mhz > policy.critical_delta_mhz:
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]: i think `--warn-delta-mhz=30` is more intuitive to read as "warn at 30 or more" (>=)?

```suggestion
    if max_delta_mhz >= policy.critical_delta_mhz:
```

add test cases to cover boundaries

File: gcm/schemas/gpu/application_clock_policy.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        severity = ClockComplianceSeverity.CRITICAL
+    elif max_delta_mhz > policy.warn_delta_mhz:
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]:
```suggestion
    elif max_delta_mhz >= policy.warn_delta_mhz:
```

File: gcm/schemas/gpu/application_clock_policy.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+from dataclasses import dataclass
+from enum import Enum
+
+from gcm.schemas.gpu.application_clock import ApplicationClockInfo
+
+
+class ClockComplianceSeverity(Enum):
+    OK = "ok"
+    WARN = "warn"
+    CRITICAL = "critical"
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]: should we do `ExitCode` directly?
```suggestion
```

File: templates/gpu_clock_policy_compliance/config-template.toml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+[health_checks.check_gpu_clock_policy]
```
Comment (luccabb) [inline templates/gpu_clock_policy_compliance/config-template.toml]: move the example to https://github.com/facebookresearch/gcm/blob/main/gcm/health_checks/config/config.toml

File: templates/gpu_clock_policy_compliance/PR-CHECKLIST.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# PR Checklist - GPU Clock Policy Compliance
```
Comment (luccabb) [inline templates/gpu_clock_policy_compliance/PR-CHECKLIST.md]: i think just remove `templates/*` dir, or add to `.gitignore`

File: gcm/schemas/gpu/README_gpu_clock_policy_implementation.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
+# GPU Clock Policy Compliance - Implementation Guide
```
Comment (luccabb) [inline gcm/schemas/gpu/README_gpu_clock_policy_implementation.md]: remove this file

File: gcm/schemas/gpu/application_clock_policy.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+def evaluate_clock_policy(
+    observed: ApplicationClockInfo, policy: ClockPolicy
+) -> ClockComplianceResult:
+    graphics_delta_mhz = abs(observed.graphics_freq - policy.expected_graphics_freq)
+    memory_delta_mhz = abs(observed.memory_freq - policy.expected_memory_freq)
+    max_delta_mhz = max(graphics_delta_mhz, memory_delta_mhz)
+
+    if max_delta_mhz >= policy.critical_delta_mhz:
+        severity = ClockComplianceSeverity.CRITICAL
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]: we've deleted `ClockComplianceSeverity`, update these to `ExitCode`
```suggestion
        severity = ExitCode.CRITICAL
```

File: gcm/schemas/gpu/application_clock_policy.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+from dataclasses import dataclass
+from enum import Enum
+
+from gcm.schemas.gpu.application_clock import ApplicationClockInfo
+
+
+
+
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]: you wanna make sure you have pre-commit installed and setup correctly such that it gives you fast feedback locally on things like: formatting, linting, typing. it'd probably delete some lines here for example instead of leaving 4 blank lines

checkout: https://github.com/facebookresearch/gcm/blob/main/gcm/CONTRIBUTING.md#environment-setup

File: website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+Validates GPU application clock compliance against a configured policy using NVML telemetry.
+For each GPU, the check compares observed graphics and memory application clocks against expected values and classifies drift as `OK`, `WARN`, or `CRITICAL`.
+
+## Command-Line Options
+
+| Option | Type | Default | Description |
+|--------|------|---------|-------------|
+| `--expected-graphics-freq` | Integer | **Required** | Expected graphics application clock (MHz) |
+| `--expected-memory-freq` | Integer | **Required** | Expected memory application clock (MHz) |
+| `--warn-delta-mhz` | Integer | 30 | Warn when absolute drift exceeds this threshold |
```
Comment (luccabb) [inline website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md]:
```suggestion
| `--warn-delta-mhz` | Integer | 30 | Warn when absolute drift meets or exceeds this threshold |
```

File: website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+| `--critical-delta-mhz` | Integer | 75 | Critical when absolute drift exceeds this threshold |
```
Comment (luccabb) [inline website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md]:
```suggestion
| `--critical-delta-mhz` | Integer | 75 | Critical when absolute drift meets or exceeds this threshold |
```

File: website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+| `--timeout` | Integer | 300 | Command execution timeout in seconds |
```
Comment (luccabb) [inline website/docs/GCM_Health_Checks/health_checks/check-gpu-clock-policy.md]: your implementation is not using `--timeout`, either add as a decorator and use where its appropriate or remove it from here

**Comment by luccabb** [comment]
@theap06 can you address all the previous comments?

File: templates/gpu_clock_policy_compliance/PR-CHECKLIST.md (hunk 6 above)
Comment (luccabb) [reply templates/gpu_clock_policy_compliance/PR-CHECKLIST.md]: @theap06 this is still applicable for example

**Comment by luccabb** [comment]
please also make sure to add a test plan to the PR summary where you're running the new check and showing it works as expected

**Comment by theap06** [comment]
> @theap06 can you address all the previous comments?

Done.

File: .gitignore [hunk 13]
```diff
@@ -29,6 +29,7 @@ out/
 debian/gcm/
 debian/.debhelper/
 .env
+templates/gpu_clock_policy_compliance/
```
Comment (luccabb) [inline .gitignore]:
```suggestion
templates/
```

then remove from the pr: 
```
git rm --cached templates/*
```

**Comment by luccabb** [comment]
for the test plan add sample runs of the new check in your machine/env:

```
(conda-env) $ <my cmd execution>
<stdout>
```

**Comment by luccabb** [comment]
@theap06 for the CI failures, can you try setting up your local env as described in https://github.com/facebookresearch/gcm/blob/main/gcm/CONTRIBUTING.md#environment-setup

specifically:

> Then run pre-commit install to install Git pre-commit hooks. You only need to do this once.

this should catch formatting/linting/typing issues before each commit, so you get fast feedback locally

File: gcm/schemas/gpu/application_clock_policy.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+from dataclasses import dataclass
+from enum import Enum
+
+from gcm.schemas.gpu.application_clock import ApplicationClockInfo
+
+
+class ClockComplianceSeverity(Enum):
+    OK = "OK"
+    WARN = "WARN"
+    CRITICAL = "CRITICAL"
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]: this is back, just use ExitCode

**Review by luccabb — CHANGES_REQUESTED** [review]

File: templates/gpu_clock_policy_compliance/PR-CHECKLIST.md (hunk 6 above)
Comment (luccabb) [reply templates/gpu_clock_policy_compliance/PR-CHECKLIST.md]: @theap06 this is still applicable, delete templates folder from the PR

File: gcm/schemas/gpu/application_clock_policy.py (hunk 14 above)
Comment (theap06) [reply gcm/schemas/gpu/application_clock_policy.py]: Done.

File: templates/gpu_clock_policy_compliance/PR-CHECKLIST.md (hunk 6 above)
Comment (luccabb) [inline templates/gpu_clock_policy_compliance/PR-CHECKLIST.md]: @theap06 delete this folder from the PR

Comment (theap06) [reply templates/gpu_clock_policy_compliance/PR-CHECKLIST.md]: Done

File: gcm/monitoring/dataclass_utils.py [hunk 15]
```diff
@@ -77,7 +77,7 @@ def asdict_recursive(obj: NonFlattened, key: str = "") -> FlattenedOrBaseType:
                 results.update(flat_result)
             else:
                 results[new_key] = flat_result
-    elif isinstance(obj, list):
+    elif isinstance(obj, list) or isinstance(obj, tuple):
```
Comment (luccabb) [inline gcm/monitoring/dataclass_utils.py]: why do we need changes in this file?

File: gcm/health_checks/checks/check_gpu_clock_policy.py [hunk 16]
```diff
@@ -0,0 +1,238 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (luccabb) [inline gcm/health_checks/checks/check_gpu_clock_policy.py]: can you move the check to be under `check_nvidia_smi`?https://github.com/facebookresearch/gcm/blob/main/gcm/health_checks/checks/check_nvidia_smi.py

move the website docs as well

**Review by luccabb — CHANGES_REQUESTED** [review]

File: gcm/monitoring/dataclass_utils.py (hunk 15 above)
Comment (theap06) [reply gcm/monitoring/dataclass_utils.py]: This is just there to check if a tuple was passed in. In hindsight, it is unnecessary here. I can remove it if you want; however, it wouldn't change any of the functionality

Comment (luccabb) [reply gcm/monitoring/dataclass_utils.py]: yeah, I think remove this for now. I'd prefer implementing this when we have a concrete use case for tuples as it's a bit unclear right now if matching tuples 1:1 with lists is actually what we'd want

File: gcm/health_checks/checks/check_gpu_clock_policy.py (hunk 16 above)
Comment (theap06) [reply gcm/health_checks/checks/check_gpu_clock_policy.py]: Done

**Comment by theap06** [comment]
I think this is good to go.

File: gcm/monitoring/dataclass_utils.py (hunk 15 above)
Comment (theap06) [reply gcm/monitoring/dataclass_utils.py]: I added the tuple implementation as the CI was failing due to that. I think it is in the test suite

Comment (luccabb) [inline gcm/monitoring/dataclass_utils.py]: remove tuple changes https://github.com/facebookresearch/gcm/pull/61/changes#r2867805937

File: gcm/tests/test_dataclasses.py [hunk 17]
```diff
@@ -199,6 +199,24 @@ class TestNestedList:
...
+    __test__ = False
+    a: tuple[tuple[str, ...], ...]
+
+
+@dataclass
+class TestTupleWithNamedObjects:
+    __test__ = False
+    a: tuple[NamedHelperDataclass, ...]
+
+
```
Comment (luccabb) [inline gcm/tests/test_dataclasses.py]: remove tuple changes https://github.com/facebookresearch/gcm/pull/61/changes#r2867805937

**Comment by theap06** [comment]
@luccabb done

File: gcm/monitoring/dataclass_utils.py [hunk 18]
```diff
@@ -90,9 +90,6 @@ def asdict_recursive(obj: NonFlattened, key: str = "") -> FlattenedOrBaseType:
                 results.update(flat_result)
             else:
                 results[new_key] = flat_result
-    elif isinstance(obj, tuple):
-        # TODO: add support for tuples
-        raise TypeError("Tuples are not supported for asdict_recursive.")
```
Comment (luccabb) [inline gcm/monitoring/dataclass_utils.py:95]: keep this file as is, this is unrelated to the new check

File: gcm/monitoring/cli/sacct_backfill.py [hunk 19]
```diff
@@ -267,7 +267,6 @@ def publish_all_parallel(
     *,
     max_concurrency: Optional[int] = None,
 ) -> None:
-    global logger
```
Comment (luccabb) [inline gcm/monitoring/cli/sacct_backfill.py:270]: same for this file, this is unrelated to the new check

File: dev-requirements.txt [hunk 20]
```diff
@@ -374,6 +374,7 @@ pynvml==11.4.1 \
     --hash=sha256:d27be542cd9d06558de18e2deffc8022ccd7355bc7382255d477038e7e424c6c
     # via gcm (pyproject.toml)
 pyoxidizer==0.24.0 \
+    --hash=sha256:1a9940d2bdb6c9e6c6c45eb4de3d4d6e5c9b3a3724dbc7d774d85d4956058447 \
```
Comment (luccabb) [inline dev-requirements.txt]:
```suggestion
```

File: gcm/monitoring/cli/sacct_backfill.py [hunk 21]
```diff
@@ -573,7 +573,6 @@ def publish_chunk(
     """
     start_iso = interval.lower.isoformat(sep="T")
     end_iso = interval.upper.isoformat(sep="T")
-    global logger
```
Comment (luccabb) [inline gcm/monitoring/cli/sacct_backfill.py:576]: keep this file as is

Comment (theap06) [reply gcm/monitoring/cli/sacct_backfill.py:576]: Do you want me to add global logger as it was previously?

Comment (luccabb) [reply gcm/monitoring/cli/sacct_backfill.py:576]: yes, I don't fully recall why this is here but believe there's a reason so no need to remove as part of this PR.

also more broadly, try to make the PRs as self contained as possible. your changes are unrelated to that file

File: gcm/health_checks/checks/check_nvidia_smi.py [hunk 22]
```diff
@@ -988,3 +1125,174 @@ def check_nvidia_smi(
 
         logger.info(f"Overall exit code {overall_exit_code}\n{overall_msg}")
         sys.exit(overall_exit_code.value)
+
```
Comment (luccabb) [inline gcm/health_checks/checks/check_nvidia_smi.py]: everything below this line already exists in the codebase, import them instead of duplicating

File: gcm/schemas/gpu/application_clock_policy.py [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+from gcm.health_checks.types import ExitCode
+from gcm.schemas.gpu.application_clock import ApplicationClockInfo
+
+
+@dataclass(frozen=True)
+class ClockPolicy:
+    expected_graphics_freq: int
+    expected_memory_freq: int
+    warn_delta_mhz: int = 0
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]:
```suggestion
    warn_delta_mhz: int
```

File: gcm/schemas/gpu/application_clock_policy.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    critical_delta_mhz: int = 0
```
Comment (luccabb) [inline gcm/schemas/gpu/application_clock_policy.py]:
```suggestion
    critical_delta_mhz: int
```

File: gcm/health_checks/checks/check_nvidia_smi.py (hunk 22 above)
Comment (theap06) [reply gcm/health_checks/checks/check_nvidia_smi.py]: Done

File: gcm/tests/health_checks_tests/test_gpu_clock_policy.py [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    tmp_path: Path,
+) -> None:
+    runner = CliRunner(mix_stderr=False)
+
+    fake_obj: NvidiaSmiCli = FakeNvidiaSmiCliObject(
+        "cluster",
+        "type",
+        "log_level",
+        "log_folder",
+        [ApplicationClockInfo(1155, 1593)],
```
Comment (luccabb) [inline gcm/tests/health_checks_tests/test_gpu_clock_policy.py]:
```suggestion
        [ApplicationClockInfo(graphics_freq=1155, memory_freq=1593)],
```

File: gcm/tests/health_checks_tests/test_gpu_clock_policy.py [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+    tmp_path: Path,
+) -> None:
+    runner = CliRunner(mix_stderr=False)
+
+    fake_obj: NvidiaSmiCli = FakeNvidiaSmiCliObject(
+        "cluster",
+        "type",
+        "log_level",
+        "log_folder",
+        [ApplicationClockInfo(1200, 1593)],
```
Comment (luccabb) [inline gcm/tests/health_checks_tests/test_gpu_clock_policy.py]:
```suggestion
        [ApplicationClockInfo(graphics_freq=1200, memory_freq=1593)],
```

File: gcm/tests/health_checks_tests/test_gpu_clock_policy.py [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+    tmp_path: Path,
+) -> None:
+    runner = CliRunner(mix_stderr=False)
+
+    fake_obj: NvidiaSmiCli = FakeNvidiaSmiCliObject(
+        "cluster",
+        "type",
+        "log_level",
+        "log_folder",
+        [ApplicationClockInfo(1300, 1593)],
```
Comment (luccabb) [inline gcm/tests/health_checks_tests/test_gpu_clock_policy.py]:
```suggestion
        [ApplicationClockInfo(graphics_freq=1300, memory_freq=1593)],
```

File: gcm/tests/health_checks_tests/test_gpu_clock_policy.py [hunk 28]
```diff
@@ -0,0 +1,187 @@
...
+
+    fake_obj: NvidiaSmiCli = FakeNvidiaSmiCliObject(
+        "cluster",
+        "type",
+        "log_level",
+        "log_folder",
+        [
+            ApplicationClockInfo(1155, 1593),
+            ApplicationClockInfo(1200, 1593),
+            ApplicationClockInfo(1300, 1593),
```
Comment (luccabb) [inline gcm/tests/health_checks_tests/test_gpu_clock_policy.py]:
```suggestion
            ApplicationClockInfo(graphics_freq=1155, memory_freq=1593),
            ApplicationClockInfo(graphics_freq=1200, memory_freq=1593),
            ApplicationClockInfo(graphics_freq=1300, memory_freq=1593),
```

File: gcm/tests/schemas/test_application_clock_policy.py [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_evaluate_clock_policy_ok() -> None:
+    policy = ClockPolicy(
+        expected_graphics_freq=1155,
+        expected_memory_freq=1593,
+        warn_delta_mhz=30,
+        critical_delta_mhz=75,
+    )
+
+    result = evaluate_clock_policy(ApplicationClockInfo(1155, 1593), policy)
```
Comment (luccabb) [inline gcm/tests/schemas/test_application_clock_policy.py]:
```suggestion
    result = evaluate_clock_policy(ApplicationClockInfo(graphics_freq=1155, memory_freq=1593), policy)
```

File: gcm/tests/schemas/test_application_clock_policy.py [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_evaluate_clock_policy_warn() -> None:
+    policy = ClockPolicy(
+        expected_graphics_freq=1155,
+        expected_memory_freq=1593,
+        warn_delta_mhz=30,
+        critical_delta_mhz=75,
+    )
+
+    result = evaluate_clock_policy(ApplicationClockInfo(1200, 1593), policy)
```
Comment (luccabb) [inline gcm/tests/schemas/test_application_clock_policy.py]:
```suggestion
    result = evaluate_clock_policy(ApplicationClockInfo(graphics_freq=1200, memory_freq=1593), policy)
```

File: gcm/tests/schemas/test_application_clock_policy.py [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+
+def test_evaluate_clock_policy_critical() -> None:
+    policy = ClockPolicy(
+        expected_graphics_freq=1155,
+        expected_memory_freq=1593,
+        warn_delta_mhz=30,
+        critical_delta_mhz=75,
+    )
+
+    result = evaluate_clock_policy(ApplicationClockInfo(1250, 1593), policy)
```
Comment (luccabb) [inline gcm/tests/schemas/test_application_clock_policy.py]:
```suggestion
    result = evaluate_clock_policy(ApplicationClockInfo(graphics_freq=1250, memory_freq=1593), policy)
```

File: gcm/health_checks/checks/check_nvidia_smi.py [hunk 32]
```diff
@@ -720,6 +794,34 @@ def process_value(self, ctx: click.Context, value: Any) -> Any:
...
+    show_default=True,
+    help="Expected GPU graphics application clock frequency (MHz).",
+)
+@click.option(
+    "--expected-memory-freq",
+    type=click.IntRange(min=0),
+    default=1593,
+    show_default=True,
+    help="Expected GPU memory application clock frequency (MHz).",
+)
```
Comment (luccabb) [inline gcm/health_checks/checks/check_nvidia_smi.py:796]: should we merge these? default to --gpu_app_freq and --gpu_app_mem_freq to keep code backwards compatible

Comment (theap06) [reply gcm/health_checks/checks/check_nvidia_smi.py:796]: done

**Comment by theap06** [comment]
I added the backwards compatability. Everything else has been addressed as well.

File: gcm/tests/schemas/test_application_clock_policy.py [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+        ClockPolicy(
+            expected_graphics_freq=1155,
+            expected_memory_freq=1593,
+            warn_delta_mhz=50,
+            critical_delta_mhz=40,
+        )
+    except ValueError:
+        return
+
+    raise AssertionError("Expected ValueError for invalid threshold ordering")
```
Comment (luccabb) [inline gcm/tests/schemas/test_application_clock_policy.py]:
```suggestion
    with pytest.raises(ValueError):
        ClockPolicy(
            expected_graphics_freq=1155,
            expected_memory_freq=1593,
            warn_delta_mhz=50,
            critical_delta_mhz=40,
        )
```

File: gcm/tests/schemas/test_application_clock_policy.py [hunk 34]
```diff
@@ -0,0 +1,73 @@
...
+def test_evaluate_clock_policy_invalid_thresholds() -> None:
+    try:
+        ClockPolicy(
+            expected_graphics_freq=1155,
+            expected_memory_freq=1593,
+            warn_delta_mhz=50,
+            critical_delta_mhz=40,
+        )
+    except ValueError:
+        return
```
Comment (luccabb) [inline gcm/tests/schemas/test_application_clock_policy.py]:
```suggestion
```

**Comment by theap06** [comment]
Why are we failing the nox style checking?

**Comment by luccabb** [comment]
pull locally and run pre-commit in the changed files

**Comment by luccabb** [comment]
ufmt is picking up unrelated changes when you run it for all files, fixing this on https://github.com/facebookresearch/gcm/pull/82
